torch/nn/parallel/replicate.py

```diff
@@ -22,15 +104,26 @@ def replicate(network, devices, detach=False):
     modules = list(network.modules())
     module_copies = [[] for device in devices]
     module_indices = {}
+    scriptmodule_skip_attr = {"_parameters", "_buffers", "_modules"}
 
     for i, module in enumerate(modules):
         module_indices[module] = i
         for j in range(num_replicas):
-            replica = module.__new__(type(module))
-            replica.__dict__ = module.__dict__.copy()
-            replica._parameters = replica._parameters.copy()
-            replica._buffers = replica._buffers.copy()
-            replica._modules = replica._modules.copy()
+            if _is_script_module(module):
+                # we have to initialize ScriptModule properly so that
+                # it works with pybind11
+                replica = _init_script_module()
+                keys = set(module.__dict__.keys()) - scriptmodule_skip_attr
+                for key in keys:
+                    replica.__dict__[key] = module.__dict__[key]
+                replica.__class__ = module.__class__
```

Comment:
I feel like this might be unsafe, since I am not sure that `__class__` uniquely determines the behavior of a class after it is constructed and there are 2 vs 3 differences in class behavior. Would `type(module).__new__()`, followed by `ScriptModule.__init__(replica)` work correctly? That should achieve the same result (constructing a member of the subclass, while initializing the ScriptModule appropriately), without having to mutate the `__class__` property.

Reply:
this is a good idea.

Reply:
I just tried the following: 

    replica = module.__new__(type(module))
    torch.jit.ScriptModule.__init__(replica)

It's coredumping. I assume it's in pybind11 related code. It's probably too hairy to go into the details there and understand why. 

Here we actually don't need to copy "__class__" attribute. That's more for my debugging purpose. I will remove this copy. The computation behavior is strictly determined by contents inside ScriptModule.

Reply:
That is true. I am curious what is going one but it is almost certainly some pybind11 magic that I don't understand.